cumulusci/tasks/salesforce/promote_2gp_package.py

```diff
@@ -0,0 +1,308 @@
+    Hence, the "promotion" of a package is just an update to the
+    Package2Version.IsReleased field.
+
+    The abbreviation `spv` in variable names is shorthand for "SubscriberPackageVersion".
+    """
+
+    task_docs = """Promote a Second Generation managed package.
+    Lists any 1GP dependencies that are detected, as well as,
+    any dependency packages that have not benn promoted.
+    Once promoted, 2GP package can be installed into production orgs."""
+
+    task_options = {
+        "version_id": {
+            "description": "The SubscriberPackageVersion (04t) Id for the target package.",
+            "required": True,
+        },
+        "auto_promote": {
+            "description": (
+                "Automatically promote any unpromoted versions of dependency 2GP packages that are detected."
+            ),
+            "required": False,
+        },
+    }
+
+    def _init_options(self, kwargs) -> None:
+        super()._init_options(kwargs)
+
+        if "version_id" not in self.options:
+            raise TaskOptionsError("Task option `version_id` is required.")
+
+        version_id = self.options["version_id"]
+        if not isinstance(version_id, str) or not version_id.startswith("04t"):
+            raise TaskOptionsError(
+                "Task option `version_id` must be a valid SubscriberPackageVersion (04t) Id"
+            )
+
+    def _init_task(self) -> None:
+        self.tooling = get_simple_salesforce_connection(
+            self.project_config,
+            get_devhub_config(self.project_config),
+            api_version=self.api_version,
+            base_url="tooling",
+        )
+
+    def _run_task(self) -> None:
+        """Orchestrate the task"""
+        version_id = self.options["version_id"]
+        dependencies = self._get_dependency_info(version_id)
+
+        self._process_one_gp_deps(dependencies)
+        should_exit = self._process_two_gp_deps(dependencies)
+
+        if should_exit:
+            return
+
+        target_package_info = self._get_target_package_info(version_id)
+        self._promote_2gp_package(target_package_info)
+
+    def _get_dependency_info(self, spv_id: str) -> Optional[List[Dict]]:
+        """
+        Given a SubscriberPackageVersionId (04t) return a list of
+        dictionaries with info on all dependency packages present.
+
+        @param spv_id: a SubscirberPackageVersionId
+        @returns: a dict with the following shape:
+        {
+            "is_2gp": (bool) whether or not this dependency is a 2GP package
+            "is_promoted": whether or not this package is already promoted (only present if is_2gp is True)
+            "Package2VersionId": The Id from the corresponding Package2Version record
+            "name": (str) The name of the package
+            "release_state": (str) the packages release state
+            "version_id": (str) the SubscriberPackageVersionId (04t) of the dependency
+        }
```

Comment:
You don't necessarily need to change this, but this looks like a situation where it might have made sense to define a pydantic model and return an instance of it instead of a dict. That way you can get help from your IDE with typing in the correct keys and assigning the correct types to them. On the other hand it's a bit more effort so maybe not worth it if we're also unit testing the code and not anticipating it changing frequently.

Reply:
I'm not sure how often we need to aggregate package information from several (related) sObjects, but I'll definitely keep this in mind next time this situation comes up. Also, looking at this again, now I want to rename the keys to match the corresponding sObject field name (e.g. {'SubscriberPackage.Name': name}).